### PR TITLE
mon,osd: limit MOSDMap messages by size as well as map count

### PR DESCRIPTION
**Pull request by liewegas**
Start with 10 MB per message.

Fixes: http://tracker.ceph.com/issues/38040

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -1469,23 +1469,53 @@ MOSDMap *OSDService::build_incremental_map_msg(epoch_t since, epoch_t to,
...
+    dout(10) << __func__ << " oldest map " << max_oldest_map << " < since "
+	     << since << ", starting with full map" << dendl;
+    since = m->oldest_map;
+    if (!get_map_bl(since, bl)) {
+      derr << __func__ << " missing full map " << since << dendl;
+      goto panic;
+    }
+    m->maps[since].claim(bl);
+    max--;
+    max_bytes -= bl.length();
```
Comment (jdurgin) [inline src/osd/OSD.cc]: isn't bl.length() 0 after it's claim()'ed above?

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -1469,23 +1469,53 @@ MOSDMap *OSDService::build_incremental_map_msg(epoch_t since, epoch_t to,
...
+  }
+  for (epoch_t e = since + 1; e <= to; ++e) {
+    bufferlist bl;
+    if (!get_inc_map_bl(e, bl)) {
+      derr << __func__ << " missing incremental map " << e << dendl;
+      goto panic;
+    }
+    m->incremental_maps[e].claim(bl);
+    max--;
+    max_bytes -= bl.length();
```
Comment (jdurgin) [inline src/osd/OSD.cc]: same here

**Comment by liewegas** [comment]
yep, fixed!

File: src/common/options.cc [hunk 3]
```diff
@@ -2578,7 +2578,11 @@ std::vector<Option> get_global_options() {
 
     Option("osd_map_message_max", Option::TYPE_INT, Option::LEVEL_ADVANCED)
     .set_default(40)
-    .set_description(""),
+    .set_description("maximum number of OSDMaps to include in a single message"),
+
+    Option("osd_map_message_max_bytes", Option::TYPE_SIZE, Option::LEVEL_ADVANCED)
+    .set_default(10_M)
```
Comment (jdurgin) [inline src/common/options.cc]: would this need be tuned for a worst-case like a large cluster w/many PGs, all upmapped?

If we get to 10M per map we'd only be sending full maps

Comment (liewegas) [reply src/common/options.cc]: a low limit just means there'd be one (incremental) map per message.

the only time we use full maps is when there is a discontinuity (client has an epoch older than our oldest map)

**Comment by liewegas** [comment]
retest this please

**Pull request merged**